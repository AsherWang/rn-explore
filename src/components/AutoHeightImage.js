import React, { useState, useEffect } from 'react';
import { Image } from 'react-native';
import { Content } from 'native-base';
import PropTypes from 'prop-types';

function AutoHeightImage({ source, customStyle }) {
  const [realHeight, setRealHeight] = useState(100);
  const [realWidth, setRealWidth] = useState(100);
  const [targetWidth, setTargetWidth] = useState(gScreen.width);


  useEffect(() => {
    const setSize = (width, height) => {
      if (width === 0 || height === 0) return;
      const targetHeight = Math.floor(targetWidth * height / width);
      setRealWidth(targetWidth);
      setRealHeight(targetHeight);
    };
    // for source with require image from local, it is a number
    if (Number.isInteger(source)) {
      const { width, height } = Image.resolveAssetSource(source);
      console.log(`get size ${width}x${height}`);
      setSize(width, height);
    } else {
      Image.getSize(source, (width, height) => {
        setSize(width, height);
      });
    }
  }, [source, targetWidth]);


  return (
    <Content
      onLayout={(event) => {
        const { width } = event.nativeEvent.layout;
        const tWidth = Math.floor(width);
        if (targetWidth !== tWidth) {
          setTargetWidth(tWidth);
        }
      }}
      style={{ ...customStyle }}
    >
      <Image
        source={source}
        style={{ height: realHeight, width: realWidth }}
      />
    </Content>
  );
}

AutoHeightImage.propTypes = {
  source: PropTypes.oneOfType([
    PropTypes.number,
    PropTypes.string,
  ]).isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  customStyle: PropTypes.object,
};

AutoHeightImage.defaultProps = {
  customStyle: {},
};

export default AutoHeightImage;
