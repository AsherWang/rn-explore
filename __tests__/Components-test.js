import 'react-native';
import React from 'react';
// import Stepper from '../src/components/Stepper';
import AutoHeightImage from '../src/components/AutoHeightImage';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

const imgSrc = require('./test.jpg');


it('renders correctly', () => {
  const tree = renderer
    .create(<AutoHeightImage source={imgSrc} style={{width:200,height:300}}/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

