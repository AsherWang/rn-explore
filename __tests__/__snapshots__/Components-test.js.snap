// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<RCTScrollView
  DEPRECATED_sendUpdatedChildFrames={false}
  alwaysBounceVertical={true}
  automaticallyAdjustContentInsets={false}
  contentContainerStyle={
    Array [
      Object {
        "padding": undefined,
      },
      undefined,
    ]
  }
  contentInset={
    Object {
      "bottom": 0,
    }
  }
  enableAutomaticScroll={true}
  enableOnAndroid={false}
  enableResetScrollToCoords={true}
  extraHeight={75}
  extraScrollHeight={0}
  getScrollResponder={[Function]}
  handleOnScroll={[Function]}
  keyboardDismissMode="interactive"
  keyboardOpeningTime={250}
  keyboardShouldPersistTaps="handled"
  keyboardSpace={0}
  onContentSizeChange={null}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onResponderGrant={[Function]}
  onResponderReject={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminationRequest={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  onScrollShouldSetResponder={[Function]}
  onStartShouldSetResponder={[Function]}
  onStartShouldSetResponderCapture={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  resetKeyboardSpace={[Function]}
  resetScrollToCoords={
    Object {
      "x": 0,
      "y": 0,
    }
  }
  scrollBarThumbImage={
    Object {
      "uri": "asset://test.png",
    }
  }
  scrollEventThrottle={1}
  scrollForExtraHeightOnAndroid={[Function]}
  scrollIntoView={[Function]}
  scrollToEnd={[Function]}
  scrollToFocusedInput={[Function]}
  scrollToPosition={[Function]}
  sendMomentumEvents={false}
  showsVerticalScrollIndicator={true}
  snapToEnd={true}
  snapToStart={true}
  style={
    Array [
      Object {
        "flexDirection": "column",
        "flexGrow": 1,
        "flexShrink": 1,
        "overflow": "scroll",
      },
      Array [
        Object {
          "backgroundColor": "transparent",
          "flex": 1,
        },
        Object {
          "height": 300,
          "width": 200,
        },
      ],
    ]
  }
  update={[Function]}
  viewIsInsideTabBar={false}
>
  <RCTScrollContentView
    collapsable={false}
    style={
      Array [
        undefined,
        Array [
          Object {
            "padding": undefined,
          },
          undefined,
        ],
      ]
    }
  >
    <Image
      source={
        Object {
          "testUri": "../../../__tests__/test.jpg",
        }
      }
      style={
        Object {
          "height": 100,
          "width": 100,
        }
      }
    />
  </RCTScrollContentView>
</RCTScrollView>
`;
